accepted/00000-jsonb-pillar-db-storage.md

```diff
@@ -0,0 +1,225 @@
++            elif self.depth == self.num_fields - 1:
+                 nk = self.num_fields-2  # Aka, self.depth-1
+                 # Should we and will we have a list at the end?
+                 if ((self.as_list and (ret[nk] in crd)) or
+```
+
+Tracked in [PR#59777](https://github.com/saltstack/salt/pull/59777).
+
+## Maintaining consistency
+
+Responsibility to provide up to date data remains, as is the case today, on the component changing the data. That means, when information about minion is changed, then relevant pillar generator must be called to refresh the pillar in database.
+However with the pillar in database and loosely (no foreign key relation) connected to minions and groups, database trigger which can remove pillar entries when minion or group is removed:
+
+```sql
+CREATE OR REPLACE FUNCTION minion_removed() RETURNS TRIGGER AS $$
+BEGIN
+  DELETE FROM suseSaltPillars WHERE target = OLD.minion_id;
+  RETURN OLD;
+END $$ LANGUAGE PLPGSQL;
+
+CREATE OR REPLACE FUNCTION group_removed() RETURNS TRIGGER AS $$
+BEGIN
+  DELETE FROM suseSaltPillars WHERE target = CONCAT('group_id:',OLD.id::VARCHAR);
+  RETURN OLD;
+END $$ LANGUAGE PLPGSQL;
+
+CREATE TRIGGER minion_removed BEFORE DELETE ON suseMinionInfo FOR EACH ROW EXECUTE PROCEDURE minion_removed();
+
+CREATE TRIGGER group_removed BEFORE DELETE ON rhnServerGroup FOR EACH ROW EXECUTE PROCEDURE group_removed();
+```
+
+In case of organization removal, enhance `delete_org` stored procedure to delete organizational pillar as well.
+
+## Performance impact
+
+I did a small scale test with 1000 minions and 10 pillar refresh repetitions with following results:
+
+Command:
+
+```bash
+time bash -c 'for i in `seq 1 100`;do salt "*" pillar.items > /dev/null; done'
+```
+
+* both suma-minion.py and DB pillars active:
+
+```
+real    12m17,561s
+user    0m49,857s
+sys     0m3,089s
+```
+
+  Connections to postgres up to 14
+
+  Postgres backends up to 7
+
+
+* only suma-minion.py pillar active:
+
+```
+real    11m11,658s
+user    0m49,353s
+sys     0m3,091s
+```
+
+
+  Connections to postgres constant 2
+
+  Postgres backends constant 2
+
+
+* only DB pillar active:
+
+```
+real    11m24,243s
+user    0m48,982s
+sys     0m3,147s
+```
+
+
+  Connections to postgres up to 14
+
+  Postgres backends up to 6
+
+
+Given the discrepancy between number of salt clients and opened connections to the database, I increased number of salt worker threads to test hypothesis that indeed number of salt threads is limiting number of connections to the database. Upon increasing salt worker threads to 28 I got following results:
+
+```
+real    11m29,790s
+user    0m47,978s
+sys     0m3,124s
+```
+
+  Connections to postgres up to 44
+
+  Postgres backends up to 14
+
+
+In conclusion I do not see very notable impact in this isolated scenario. Number of opened connections is correlated with number of salt worker threads, this needs to be taking into account in case of tuning individual installations.
```

Comment:
Remark for implementation time: update "The LSD" https://github.com/uyuni-project/uyuni-docs/blob/master/modules/large-deployments/pages/tuning.adoc

I can help with that, just ping me when the time comes